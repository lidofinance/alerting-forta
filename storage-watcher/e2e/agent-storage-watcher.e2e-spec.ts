import { configureContainer, Finding } from "forta-agent";
import { AwilixContainer, asFunction } from "awilix";
import {
  provideAgentPath,
  provideRunBlock,
  provideRunTransaction,
} from "./utils";

const TEST_TIMEOUT = 60_000; // ms

describe("agent-storage-watcher e2e tests", () => {
  let runBlock: (
    blockHashOrNumber: string | number,
    initBlock?: number,
  ) => Promise<Finding[]>;
  let runTransaction: (txHash: string) => Promise<Finding[]>;
  let logSpy: jest.SpyInstance;
  let timeSpy: jest.SpyInstance;

  beforeAll(() => {
    logSpy = jest.spyOn(console, "log");
    logSpy.mockImplementation(() => {});
    timeSpy = jest.spyOn(Date, "now");
    timeSpy.mockImplementation(() => new Date("2023-12-31"));
  });

  beforeEach(async () => {
    const container = configureContainer() as AwilixContainer;
    container.register({
      agentPath: asFunction(
        provideAgentPath("subagents/storage-watcher/agent-storage-watcher"),
      ),
      runTransaction: asFunction(provideRunTransaction),
      runBlock: asFunction(provideRunBlock),
    });

    // https://docs.forta.network/en/latest/cli/#invoke-commands-programmatically
    runTransaction = container.resolve("runTransaction");
    runBlock = container.resolve("runBlock");
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it(
    "should process block with critical storage slot value changed",
    async () => {
      const findings = await runBlock(16398431, 16398430);
      expect(findings).toMatchSnapshot();
    },
    TEST_TIMEOUT,
  );
});
