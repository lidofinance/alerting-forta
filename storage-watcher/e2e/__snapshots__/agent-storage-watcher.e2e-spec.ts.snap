// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`agent-storage-watcher e2e tests should process block with critical storage slot value changed 1`] = `
Array [
  Finding {
    "addresses": Array [],
    "alertId": "STORAGE-SLOT-VALUE-CHANGED",
    "description": "Value of the storage slot 'manager' for contract 0xF95f069F9AD107938F6ba802a3da87892298610E (MEV Boost Relay Allowed List) has changed!
Prev value: 0x0000000000000000000000000000000000000000000000000000000000000000
New value: 0x00000000000000000000000098be4a407bff0c125e25fbe9eb1165504349c37d",
    "labels": Array [],
    "metadata": Object {},
    "name": "ðŸš¨ Critical storage slot value changed",
    "protocol": "ethereum",
    "severity": 4,
    "type": 2,
  },
]
`;
